sql/src/main/java/io/crate/lucene/ArrayLengthQuery.java

```diff
@@ -0,0 +1,413 @@
+        MappedFieldType fieldType = context.getFieldTypeOrNull(arrayRef.column().fqn());
+        if (fieldType == null) {
+            return Queries.newMatchNoDocsQuery("MappedFieldType missing");
+        }
+        return fieldType.existsQuery(context.queryShardContext());
+    }
+
+    private static IntPredicate predicateForFunction(String cmpFuncName, int cmpValue) {
+        switch (cmpFuncName) {
+            case LtOperator.NAME:
+                return x -> x < cmpValue;
+
+            case LteOperator.NAME:
+                return x -> x <= cmpValue;
+
+            case GtOperator.NAME:
+                return x -> x > cmpValue;
+
+            case GteOperator.NAME:
+                return x -> x >= cmpValue;
+
+            case EqOperator.NAME:
+                return x -> x == cmpValue;
+
+            default:
+                throw new IllegalArgumentException("Unknown comparison function: " + cmpFuncName);
+        }
+    }
+
+    private static IntUnaryOperator getNumTermsPerDocFunction(LeafReader reader, Reference ref) {
+        DataType elementType = CollectionType.unnest(ref.valueType());
+        switch (elementType.id()) {
+            case BooleanType.ID:
+                return numValuesPerDocForBoolean(reader, ref.column());
+
+            case ByteType.ID:
+            case ShortType.ID:
+            case IntegerType.ID:
+            case LongType.ID:
+            case TimestampType.ID:
+            case FloatType.ID:
+            case DoubleType.ID:
+            case GeoPointType.ID:
+                return numValuesPerDocForSortedNumeric(reader, ref.column());
+
+            case StringType.ID:
+                return numValuesPerDocForString(reader, ref.column());
+
+            case IpType.ID:
+                return numValuesPerDocForIP(reader, ref.column());
+
+            default:
+                throw new UnsupportedOperationException("NYI: " + elementType);
+        }
+    }
+
+    private static IntUnaryOperator numValuesPerDocForIP(LeafReader reader, ColumnIdent column) {
+        SortedSetDocValues docValues;
+        try {
+            docValues = reader.getSortedSetDocValues(column.fqn());
+        } catch (IOException e) {
+            throw new RuntimeException(e);
+        }
+        return doc -> {
+            try {
+                if (docValues.advanceExact(doc)) {
+                    int c = 1;
+                    while (docValues.nextOrd() != SortedSetDocValues.NO_MORE_ORDS) {
+                        c++;
+                    }
+                    return c;
+                }
+            } catch (IOException e) {
+                throw new RuntimeException(e);
+            }
+            return 0;
+        };
+    }
+
+    private static IntUnaryOperator numValuesPerDocForBoolean(LeafReader reader, ColumnIdent column) {
+        SortedBinaryDocValues docValues;
+        try {
+            docValues = FieldData.toString(DocValues.getSortedNumeric(reader, column.fqn()));
```

Comment:
isn't it irrelevant to wrap it into `toString` as long as we are only interested in the value count?